<a href="https://colab.research.google.com/github/qurban-12/ML_internship_Qurban/blob/main/work/notebooks/w05_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-08 — Capstone Modeling Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/qurban-12/ML_internship_Qurban/blob/main/work/notebooks/w05_model.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Method choice and why

*Which method from the toolkit, and why it fits your lane.*

I will start with Logistic Regression because the target is a binary observed label: whether a content item is declining or not. It gives a simple and interpretable baseline for this modeling task.

I will use the official 26 model features from the project feature list and avoid `trend_direction` and `trend_pct` because they are used to create the target label. I will then compare the model with the Week-4 baseline using the same data, grouped split, and Precision@K metric.

If useful, I will also test a Random Forest as a second model and compare whether the added complexity gives a meaningful improvement.


In [35]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# ML-08 — Train and Compare the Model

# ML-08 — Train and Compare the Model

import pandas as pd
import numpy as np

from pathlib import Path

RANDOM_STATE = 42

DATA_PATH = Path("/content/ML_internship_Qurban/data/raw/content_refresh_anonymized.csv")

df = pd.read_csv(DATA_PATH)

print("Dataset shape:", df.shape)
print("Columns:", len(df.columns))

df.head()

Dataset shape: (30000, 44)
Columns: 44


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [36]:
import os

print("Current folder:")
print(os.getcwd())

print("\nFiles/folders in /content:")
print(os.listdir("/content"))

Current folder:
/content/ML_internship_Qurban

Files/folders in /content:
['.config', 'ML_internship_Qurban', 'sample_data']


In [37]:
!git clone https://github.com/qurban-12/ML_internship_Qurban.git

Cloning into 'ML_internship_Qurban'...
remote: Enumerating objects: 151, done.
remote: Counting objects: 100% (151/151), done.
remote: Compressing objects: 100% (107/107), done.
remote: Total 151 (delta 57), reused 95 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (151/151), 1.89 MiB | 11.05 MiB/s, done.
Resolving deltas: 100% (57/57), done.


In [38]:
import os

csv_path = "/content/ML_internship_Qurban/data/raw/content_refresh_anonymized.csv"

print("CSV exists:", os.path.exists(csv_path))
print("CSV path:", csv_path)

CSV exists: True
CSV path: /content/ML_internship_Qurban/data/raw/content_refresh_anonymized.csv


In [39]:
import pandas as pd

DATA_PATH = "/content/ML_internship_Qurban/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Dataset shape:", df.shape)
print("Columns:", len(df.columns))

df.head()

Dataset shape: (30000, 44)
Columns: 44


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [40]:
from pathlib import Path

prep_path = Path("/content/ML_internship_Qurban/scripts/01_prepare_features.py")

print("File exists:", prep_path.exists())
print("Path:", prep_path)


File exists: True
Path: /content/ML_internship_Qurban/scripts/01_prepare_features.py


In [41]:
print(prep_path.read_text())

from __future__ import annotations

import argparse
from pathlib import Path

import numpy as np
import pandas as pd

from ml_utils import (
    BOOLEAN_COLUMNS,
    CATEGORICAL_COLUMNS,
    MODEL_CATEGORICAL_FEATURES,
    MODEL_NUMERIC_FEATURES,
    NUMERIC_COLUMNS,
    PROCESSED_DIR,
    RAW_PATH,
    display_path,
    ensure_dirs,
    to_bool_series,
    write_json,
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Prepare FlyRank refresh feature vector.")
    parser.add_argument("--input", default=str(RAW_PATH), help="Raw anonymized CSV export.")
    parser.add_argument(
        "--output",
        default=str(PROCESSED_DIR / "refresh_feature_vector.csv"),
        help="Prepared feature-vector CSV.",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    ensure_dirs()

    input_path = Path(args.input)
    if not input_path.exists():
        raise FileNotFoundError(
            f"Raw input not found: {input

In [42]:
%cd /content/ML_internship_Qurban

!python scripts/01_prepare_features.py

/content/ML_internship_Qurban
Prepared 30,000 rows from 30,000 raw rows
Wrote /content/ML_internship_Qurban/data/processed/refresh_feature_vector.csv


In [43]:
import pandas as pd

PREPARED_PATH = "/content/ML_internship_Qurban/data/processed/refresh_feature_vector.csv"

df = pd.read_csv(PREPARED_PATH)

print("Prepared dataset shape:", df.shape)
print("Number of columns:", len(df.columns))

print("\nTarget distribution:")
print(df["is_declining_label"].value_counts())

print("\nTarget percentage:")
print(
    df["is_declining_label"]
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
)

Prepared dataset shape: (30000, 52)
Number of columns: 52

Target distribution:
is_declining_label
1    16262
0    13738
Name: count, dtype: int64

Target percentage:
is_declining_label
1    54.21
0    45.79
Name: proportion, dtype: float64


In [44]:
from pathlib import Path

utils_path = Path("/content/ML_internship_Qurban/scripts/ml_utils.py")

print("File exists:", utils_path.exists())

File exists: True


In [45]:
print(utils_path.read_text())

from __future__ import annotations

import json
from pathlib import Path
from typing import Iterable

import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
RAW_PATH = ROOT / "data" / "raw" / "content_refresh_anonymized.csv"
PROCESSED_DIR = ROOT / "data" / "processed"
OUTPUT_DIR = ROOT / "outputs"
CHART_DIR = OUTPUT_DIR / "charts"


NUMERIC_COLUMNS = [
    "search_volume",
    "competition",
    "cpc",
    "word_count",
    "char_count",
    "impressions_90d",
    "clicks_90d",
    "pageviews_90d",
    "sessions_90d",
    "users_90d",
    "engaged_sessions_90d",
    "ai_sessions_90d",
    "scroll_events_90d",
    "days_with_impressions",
    "days_with_sessions",
    "impressions_last_30d",
    "clicks_last_30d",
    "sessions_last_30d",
    "impressions_prev_30d",
    "clicks_prev_30d",
    "sessions_prev_30d",
    "content_age_days",
    "age_tier_order",
    "days_since_last_update",
    "ctr",
    "avg_position",
    "engagement_rate",
    "scroll_rate"

In [46]:
# Official model features from scripts/ml_utils.py

TARGET = "is_declining_label"
GROUP_COLUMN = "client_id"

MODEL_NUMERIC_FEATURES = [
    "search_volume",
    "competition",
    "cpc",
    "word_count",
    "char_count",
    "log_impressions_90d",
    "log_clicks_90d",
    "log_sessions_90d",
    "log_ai_sessions_90d",
    "days_with_impressions",
    "days_with_sessions",
    "content_age_days",
    "days_since_last_update",
    "ctr",
    "avg_position",
    "engagement_rate",
    "scroll_rate",
    "ai_traffic_pct",
]

MODEL_CATEGORICAL_FEATURES = [
    "competition_level",
    "content_type",
    "main_intent",
    "age_tier",
    "freshness_tier",
    "word_count_tier",
    "impression_tier",
    "position_tier",
]

print("Numeric features:", len(MODEL_NUMERIC_FEATURES))
print("Categorical features:", len(MODEL_CATEGORICAL_FEATURES))
print("Total model features:", len(MODEL_NUMERIC_FEATURES) + len(MODEL_CATEGORICAL_FEATURES))

Numeric features: 18
Categorical features: 8
Total model features: 26


In [47]:
# Check missing values in official model features

missing_summary = (
    df[MODEL_NUMERIC_FEATURES + MODEL_CATEGORICAL_FEATURES]
    .isna()
    .sum()
    .sort_values(ascending=False)
)

print(missing_summary)

search_volume             0
competition               0
cpc                       0
word_count                0
char_count                0
log_impressions_90d       0
log_clicks_90d            0
log_sessions_90d          0
log_ai_sessions_90d       0
days_with_impressions     0
days_with_sessions        0
content_age_days          0
days_since_last_update    0
ctr                       0
avg_position              0
engagement_rate           0
scroll_rate               0
ai_traffic_pct            0
competition_level         0
content_type              0
main_intent               0
age_tier                  0
freshness_tier            0
word_count_tier           0
impression_tier           0
position_tier             0
dtype: int64


In [48]:
# Check client groups

print("Number of clients:", df["client_id"].nunique())

print("\nRows per client:")
print(df["client_id"].value_counts().describe())

Number of clients: 32

Rows per client:
count      32.000000
mean      937.500000
std      1376.387113
min         3.000000
25%       110.250000
50%       567.000000
75%      1058.750000
max      7008.000000
Name: count, dtype: float64


## 2. Split design

*Grouped by client? Time-aware? Say why this split is honest for your question.*

I will use a grouped train/test split by `client_id`. This keeps all content from the same client in only one side of the split, which reduces the risk of learning client-specific patterns.

I will use 80% of the data for training and 20% for testing, with a fixed random seed for reproducibility. The same test split will be used to evaluate both the Week-4 baseline and the new model.


In [49]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
from sklearn.model_selection import GroupShuffleSplit

# Grouped train/test split by client
X = df[MODEL_NUMERIC_FEATURES + MODEL_CATEGORICAL_FEATURES].copy()
y = df[TARGET].copy()
groups = df[GROUP_COLUMN].copy()

splitter = GroupShuffleSplit(
    n_splits=1,
    test_size=0.20,
    random_state=42
)

train_idx, test_idx = next(
    splitter.split(X, y, groups=groups)
)

X_train = X.iloc[train_idx].copy()
X_test = X.iloc[test_idx].copy()

y_train = y.iloc[train_idx].copy()
y_test = y.iloc[test_idx].copy()

groups_train = groups.iloc[train_idx].copy()
groups_test = groups.iloc[test_idx].copy()

print("Training rows:", len(X_train))
print("Testing rows:", len(X_test))

print("\nTraining clients:", groups_train.nunique())
print("Testing clients:", groups_test.nunique())

print("\nClients shared between train and test:",
      len(set(groups_train) & set(groups_test)))

print("\nTraining target rate:", round(y_train.mean() * 100, 2), "%")
print("Testing target rate:", round(y_test.mean() * 100, 2), "%")

Training rows: 23837
Testing rows: 6163

Training clients: 25
Testing clients: 7

Clients shared between train and test: 0

Training target rate: 55.01 %
Testing target rate: 51.1 %


## 3. Train + compare vs my baseline

*Same data, same metric, same split as your Week-4 baseline. Show the table.*

In [50]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression

# Preprocessing for numeric features
numeric_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

# Preprocessing for categorical features
categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer([
    ("numeric", numeric_pipeline, MODEL_NUMERIC_FEATURES),
    ("categorical", categorical_pipeline, MODEL_CATEGORICAL_FEATURES)
])

# Logistic Regression model
logistic_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", LogisticRegression(
        max_iter=1000,
        random_state=42
    ))
])

# Train
logistic_model.fit(X_train, y_train)

# Predicted probability of declining
logistic_scores = logistic_model.predict_proba(X_test)[:, 1]

print("Logistic Regression trained successfully.")
print("Number of test predictions:", len(logistic_scores))
print("Minimum predicted score:", round(logistic_scores.min(), 4))
print("Maximum predicted score:", round(logistic_scores.max(), 4))

Logistic Regression trained successfully.
Number of test predictions: 6163
Minimum predicted score: 0.0284
Maximum predicted score: 0.9574


In [51]:
# Recreate the Week-4 baseline rule

STALE_DAYS = 91
GOOD_POSITION_MAX = 10
LOW_CTR_MAX = 0.10

baseline_score_all = (
    (df["days_since_last_update"] >= STALE_DAYS).astype(int) * 2
    +
    (
        (df["avg_position"] >= 1)
        & (df["avg_position"] <= GOOD_POSITION_MAX)
        & (df["ctr"] < LOW_CTR_MAX)
    ).astype(int)
)

# Evaluate baseline on the exact same test rows
baseline_scores_test = baseline_score_all.iloc[test_idx].to_numpy()

print("Baseline scores on test set:")
print(pd.Series(baseline_scores_test).value_counts().sort_index())

print("\nTest baseline scores:", len(baseline_scores_test))
print("Test labels:", len(y_test))

Baseline scores on test set:
0    3650
1    1295
2     940
3     278
Name: count, dtype: int64

Test baseline scores: 6163
Test labels: 6163


I trained a Logistic Regression model using the official 26 features and evaluated it on the same client-grouped test set as the Week-4 baseline.

On the test set, the base rate of declining content was 51.1%. The Week-4 baseline and Logistic Regression both measured 0.700 Precision@20. Logistic Regression measured 0.720 Precision@50 versus 0.620 for the baseline, and 0.700 Precision@100 versus 0.550 for the baseline.

These results are directional decision-support measurements on this test split. They show that Logistic Regression provides stronger precision at K=50 and K=100, while the two approaches were equal at K=20.


In [52]:
from scripts.ml_utils import precision_at_k

# Precision@K comparison
K_VALUES = [20, 50, 100]

comparison_rows = []

for k in K_VALUES:
    comparison_rows.append({
        "Method": "Week-4 Baseline",
        "Precision@K": f"Precision@{k}",
        "Value": precision_at_k(y_test, baseline_scores_test, k)
    })

    comparison_rows.append({
        "Method": "Logistic Regression",
        "Precision@K": f"Precision@{k}",
        "Value": precision_at_k(y_test, logistic_scores, k)
    })

comparison_df = pd.DataFrame(comparison_rows)

print("Test-set base rate:",
      round(y_test.mean() * 100, 2), "%")

print("\nComparison:")
display(
    comparison_df.style.format({"Value": "{:.3f}"})
)

Test-set base rate: 51.1 %

Comparison:


,Method,Precision@K,Value
0,Week-4 Baseline,Precision@20,0.700
1,Logistic Regression,Precision@20,0.700
2,Week-4 Baseline,Precision@50,0.620
3,Logistic Regression,Precision@50,0.720
4,Week-4 Baseline,Precision@100,0.550
5,Logistic Regression,Precision@100,0.700


In [53]:
from sklearn.metrics import roc_auc_score, average_precision_score

logistic_roc_auc = roc_auc_score(y_test, logistic_scores)
logistic_pr_auc = average_precision_score(y_test, logistic_scores)

print("Logistic Regression ROC-AUC:", round(logistic_roc_auc, 3))
print("Logistic Regression Average Precision:", round(logistic_pr_auc, 3))

Logistic Regression ROC-AUC: 0.616
Logistic Regression Average Precision: 0.604


In [54]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 4. Errors and interpretation

*Where is the model wrong? What does it lean on? A short error analysis beats a big metric table.*

The model is useful for ranking declining content, but its predictions are not perfect. I will inspect classification errors and feature importance to understand where the model is making mistakes and what signals it relies on.

I will use permutation importance as a diagnostic. A feature with higher importance means that changing that feature affected the model's measured performance more on the test set. This shows association with the model's predictions, not a causal relationship.


In [55]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
from sklearn.metrics import confusion_matrix, classification_report

# Convert probabilities to 0/1 predictions using a 0.5 threshold
logistic_predictions = (logistic_scores >= 0.5).astype(int)

# Confusion matrix
cm = confusion_matrix(y_test, logistic_predictions)

print("Confusion Matrix:")
print(cm)

print("\nClassification Report:")
print(
    classification_report(
        y_test,
        logistic_predictions,
        target_names=["Not declining", "Declining"],
        digits=3
    )
)

Confusion Matrix:
[[1339 1675]
 [ 868 2281]]

Classification Report:
               precision    recall  f1-score   support

Not declining      0.607     0.444     0.513      3014
    Declining      0.577     0.724     0.642      3149

     accuracy                          0.587      6163
    macro avg      0.592     0.584     0.578      6163
 weighted avg      0.591     0.587     0.579      6163



The Logistic Regression model correctly identified 2,281 of the 3,149 declining items, giving a recall of 0.724 for the declining class. However, it also predicted 1,675 non-declining items as declining, which are false positives.

Overall accuracy was 0.587, so the model should not be treated as a perfect classifier. For this task, ranking quality is more relevant than a fixed 0.5 classification threshold because the main goal is to identify useful items for review. I will therefore inspect concrete errors and feature importance before drawing conclusions about the model.


In [56]:
# Build an error-analysis dataframe from the test set
error_df = df.iloc[test_idx].copy()

error_df["actual"] = y_test.to_numpy()
error_df["predicted"] = logistic_predictions
error_df["prediction_score"] = logistic_scores

# False positives: predicted declining, but actually not declining
false_positives = error_df[
    (error_df["actual"] == 0) &
    (error_df["predicted"] == 1)
].copy()

# False negatives: actually declining, but predicted not declining
false_negatives = error_df[
    (error_df["actual"] == 1) &
    (error_df["predicted"] == 0)
].copy()

print("False positives:", len(false_positives))
print("False negatives:", len(false_negatives))

print("\nExample false positives:")
display(
    false_positives[
        [
            "content_id",
            "content_type",
            "days_since_last_update",
            "avg_position",
            "ctr",
            "prediction_score",
            "actual",
            "predicted"
        ]
    ].head(3)
)

print("\nExample false negatives:")
display(
    false_negatives[
        [
            "content_id",
            "content_type",
            "days_since_last_update",
            "avg_position",
            "ctr",
            "prediction_score",
            "actual",
            "predicted"
        ]
    ].head(3)
)

False positives: 1675
False negatives: 868

Example false positives:


,content_id,content_type,days_since_last_update,avg_position,ctr,prediction_score,actual,predicted
13,content_a5a2fbc76336,keyword article,103,39.8,0.00,0.737550,0,1
26,content_72c5c2d73e5a,keyword article,13,30.0,0.12,0.628169,0,1
36,content_bce275871a25,keyword article,20,5.4,1.35,0.544712,0,1



Example false negatives:


,content_id,content_type,days_since_last_update,avg_position,ctr,prediction_score,actual,predicted
39,content_4595e8704e07,keyword article,104,36.3,0.0,0.209629,1,0
47,content_40cb4af260c0,keyword article,20,25.5,0.0,0.381569,1,0
58,content_caff51984338,keyword article,20,9.1,0.0,0.404375,1,0


The concrete errors show why the model should be used as decision support rather than as a definitive classifier. Some false positives have signals associated with decline, such as older content or very low CTR, but their observed label is not declining. For example, one false positive was 103 days since its last update with a CTR of 0.00%, while another had been updated only 13 days earlier.

The false negatives show the opposite problem. Some items were actually declining but received relatively low predicted probabilities. For example, one false negative was 104 days since its last update with a CTR of 0.00%, but the model assigned it a probability of only 0.210.

These examples suggest that individual signals do not fully determine the observed outcome. This is one reason to evaluate the model as a ranking system using Precision@K rather than relying only on a 0.5 classification threshold.


In [57]:
from sklearn.inspection import permutation_importance

# Measure how much each original feature affects model performance
perm = permutation_importance(
    logistic_model,
    X_test,
    y_test,
    scoring="roc_auc",
    n_repeats=10,
    random_state=42,
    n_jobs=-1
)

importance_df = pd.DataFrame({
    "feature": X_test.columns,
    "importance_mean": perm.importances_mean,
    "importance_std": perm.importances_std
}).sort_values(
    "importance_mean",
    ascending=False
)

print("Top 10 features by permutation importance:")

display(
    importance_df.head(10).style.format({
        "importance_mean": "{:.4f}",
        "importance_std": "{:.4f}"
    })
)

Top 10 features by permutation importance:


,feature,importance_mean,importance_std
5,log_impressions_90d,0.0803,0.0057
6,log_clicks_90d,0.0581,0.0030
11,content_age_days,0.0342,0.0028
24,impression_tier,0.0337,0.0037
14,avg_position,0.0256,0.0030
7,log_sessions_90d,0.0236,0.0013
3,word_count,0.0096,0.0022
10,days_with_sessions,0.0073,0.0007
16,scroll_rate,0.0063,0.0022
18,competition_level,0.0048,0.0010


Permutation importance shows that `log_impressions_90d` was the most influential feature for the Logistic Regression model, with a mean importance of 0.0803. It was followed by `log_clicks_90d` at 0.0581 and `content_age_days` at 0.0342.

Other useful signals included `impression_tier`, `avg_position`, and `log_sessions_90d`. These results suggest that the model relies strongly on recent traffic and visibility-related measurements, together with content age and search-position information.

The importance values describe how much the features contribute to the model's measured predictive performance. They should not be interpreted as causal effects. The results also provide a useful sanity check because the important features are related to content visibility, traffic, age, and position rather than the target-defining fields `trend_direction` or `trend_pct`.


In [58]:
# Final feature leakage check
forbidden_features = {
    "trend_direction",
    "trend_pct",
    "is_declining_label",
    "client_id",
    "content_id"
}

used_features = set(
    MODEL_NUMERIC_FEATURES + MODEL_CATEGORICAL_FEATURES
)

leakage_features = sorted(
    used_features & forbidden_features
)

print("Number of model features:", len(used_features))
print("Forbidden features used:", leakage_features)

if len(leakage_features) == 0:
    print("Leakage check passed: no forbidden target/ID features are used.")
else:
    print("WARNING: forbidden features detected.")

Number of model features: 26
Forbidden features used: []
Leakage check passed: no forbidden target/ID features are used.


In [59]:
from sklearn.ensemble import RandomForestClassifier

# Random Forest does not need feature scaling
rf_preprocessor = ColumnTransformer([
    ("numeric", SimpleImputer(strategy="median"), MODEL_NUMERIC_FEATURES),
    (
        "categorical",
        Pipeline([
            ("imputer", SimpleImputer(strategy="most_frequent")),
            ("onehot", OneHotEncoder(handle_unknown="ignore"))
        ]),
        MODEL_CATEGORICAL_FEATURES
    )
])

random_forest_model = Pipeline([
    ("preprocessor", rf_preprocessor),
    ("model", RandomForestClassifier(
        n_estimators=300,
        min_samples_leaf=5,
        random_state=42,
        n_jobs=-1
    ))
])

random_forest_model.fit(X_train, y_train)

rf_scores = random_forest_model.predict_proba(X_test)[:, 1]

print("Random Forest trained successfully.")
print("Number of test predictions:", len(rf_scores))
print("Minimum predicted score:", round(rf_scores.min(), 4))
print("Maximum predicted score:", round(rf_scores.max(), 4))

Random Forest trained successfully.
Number of test predictions: 6163
Minimum predicted score: 0.0297
Maximum predicted score: 0.9583


In [60]:
# Compare all methods on the same test set
final_comparison_rows = []

for k in K_VALUES:
    final_comparison_rows.append({
        "Method": "Week-4 Baseline",
        "Metric": f"Precision@{k}",
        "Value": precision_at_k(
            y_test,
            baseline_scores_test,
            k
        )
    })

    final_comparison_rows.append({
        "Method": "Logistic Regression",
        "Metric": f"Precision@{k}",
        "Value": precision_at_k(
            y_test,
            logistic_scores,
            k
        )
    })

    final_comparison_rows.append({
        "Method": "Random Forest",
        "Metric": f"Precision@{k}",
        "Value": precision_at_k(
            y_test,
            rf_scores,
            k
        )
    })

final_comparison_df = pd.DataFrame(final_comparison_rows)

print("Test-set base rate:",
      round(y_test.mean() * 100, 2), "%")

print("\nFinal comparison:")
display(
    final_comparison_df.style.format({
        "Value": "{:.3f}"
    })
)

Test-set base rate: 51.1 %

Final comparison:


,Method,Metric,Value
0,Week-4 Baseline,Precision@20,0.700
1,Logistic Regression,Precision@20,0.700
2,Random Forest,Precision@20,0.750
3,Week-4 Baseline,Precision@50,0.620
4,Logistic Regression,Precision@50,0.720
5,Random Forest,Precision@50,0.680
6,Week-4 Baseline,Precision@100,0.550
7,Logistic Regression,Precision@100,0.700
8,Random Forest,Precision@100,0.700


The three approaches were evaluated on the same client-grouped test set using the same Precision@K metric. The test-set base rate of declining content was 51.1%.

At Precision@20, the Week-4 baseline and Logistic Regression both achieved 0.700, while Random Forest achieved 0.750. At Precision@50, Logistic Regression achieved 0.720 compared with 0.680 for Random Forest and 0.620 for the Week-4 baseline. At Precision@100, Logistic Regression and Random Forest both achieved 0.700, compared with 0.550 for the baseline.

The results show that model performance depends on the size of the review set. Random Forest produced the highest measured Precision@20, while Logistic Regression produced higher measured Precision@50 and the same Precision@100 as Random Forest. Therefore, the additional complexity of Random Forest did not provide a consistent improvement across all review depths.

For this task, Precision@K is more useful than accuracy alone because the workflow is intended to prioritize a limited number of content items for review. These results should be treated as directional evidence from this client-grouped test split rather than as proof that one model will perform the same way on future data.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

### Self-check

* **Same split and baseline:** The Week-4 baseline, Logistic Regression, and Random Forest were evaluated on the same client-grouped test split.
* **Valid validation design:** The split used `client_id` groups, with no client appearing in both training and testing data.
* **Method choice explained:** Logistic Regression was selected first because the target is a binary observed label. Random Forest was added to test whether a more complex model provides additional value.
* **Useful metrics:** Precision@20, Precision@50, and Precision@100 were used because the task is to rank content items for review. ROC-AUC and Average Precision were also calculated for Logistic Regression as supporting diagnostics.
* **Errors were inspected:** False positives and false negatives were reviewed using concrete examples.
* **Feature interpretation:** Permutation importance was used to identify influential features. The strongest signals included impressions, clicks, content age, impression tier, position, and sessions.
* **Leakage check:** The final feature check found no forbidden target or ID features among the 26 model features.
* **Complexity check:** Random Forest did not improve consistently across all Precision@K levels, so added model complexity was not treated as valuable by itself.
* **Reproducibility:** Fixed random seeds were used for the grouped split, Logistic Regression, Random Forest, and permutation-importance analysis.
